### PR TITLE
Feature/bookmarks

### PR DESCRIPTION
**Pull request by fominmaksim**
- created rest api for tweets bookmarks

### CODE REVIEW

File: src/models/Bookmark.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import mongoose, { model, Schema, Document } from "mongoose";
+
+export interface Bookmark extends Document {
+  tweetId: ObjectId;
+  ownerId: ObjectId;
+}
+
+const BookmarkSchema: Schema = new Schema<Bookmark>({
+  tweetId: {
+    type: mongoose.Schema.Types.ObjectId,
```
Comment (VitaliiDenysiuk) [inline src/models/Bookmark.ts]: why did you use type "type: mongoose.Schema.Types.ObjectId", instead "type: ObjectId"?

File: src/repositories/bookmarks.repository.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import { BookmarkModel } from "../models/Bookmark";
+
+export const listBookmarks = async () => {
+  const result = await BookmarkModel.find();
+  return result;
+};
+
+export const addOne = async (tweetId: ObjectId, ownerId: ObjectId) => {
+  const result = new BookmarkModel({
```
Comment (VitaliiDenysiuk) [inline src/repositories/bookmarks.repository.ts:8]: In that place we need to string-type parameters of function. And inside function take ObjectId by string. Only repository has to know how works with database.

File: src/repositories/bookmarks.repository.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    tweetId: tweetId,
+    ownerId: ownerId,
+  });
+  return result.save();
+};
+
+export const deleteOne = async (id: ObjectId) => {
```
Comment (VitaliiDenysiuk) [inline src/repositories/bookmarks.repository.ts]: same

File: src/models/Bookmark.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { ObjectId } from "mongodb";
```
Comment (VitaliiDenysiuk) [inline src/models/Bookmark.ts]: rename file Bookmark.model.ts

File: src/services/bookmarks.service.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { ObjectId } from "mongodb";
```
Comment (VitaliiDenysiuk) [inline src/services/bookmarks.service.ts]: did not use database inside services. 
Only the repository needs to know which database we are working with.

File: src/services/bookmarks.service.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+export const getBookmarks = async () => {
+  const bookmarksList: Bookmark[] = await listBookmarks();
+  return bookmarksList;
+};
+
+export const addBookmark = async (ownerId: string, tweetId: string) => {
+  if (!ownerId || !tweetId) {
+    throw new Error("ownerId and tweetId are required");
+  } else {
+    return addOne(new ObjectId(ownerId), new ObjectId(tweetId));
```
Comment (VitaliiDenysiuk) [inline src/services/bookmarks.service.ts]: send string parameters and in the repository get new ObjectId(ownerId)

File: src/services/bookmarks.service.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  }
+};
+
+export const deleteBookmark = (id: string) => {
+  return deleteOne(new ObjectId(id));
```
Comment (VitaliiDenysiuk) [inline src/services/bookmarks.service.ts]: send string parameters and in the repository get new ObjectId(ownerId)

**Review by VitaliiDenysiuk — CHANGES_REQUESTED** [review]

File: src/models/Bookmark.ts (hunk 4 above)
Comment (fominmaksim) [reply src/models/Bookmark.ts]: I did. but why does User.ts isn't named like this?

File: src/models/Bookmark.model.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import { ObjectId } from "mongodb";
+import mongoose, { model, Schema, Document } from "mongoose";
+
+export interface Bookmark extends Document {
+  tweetId: ObjectId;
+  ownerId: ObjectId;
```
Comment (alexandrtovmach) [inline src/models/Bookmark.model.ts:6]: createdAt, updatedAt

File: src/models/Bookmark.model.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+const BookmarkSchema: Schema = new Schema<Bookmark>({
+  tweetId: {
+    type: mongoose.Schema.Types.ObjectId,
+    required: true,
+    ref: "Tweet",
+  },
+
+  ownerId: {
+    type: ObjectId,
+    required: true,
```
Comment (alexandrtovmach) [inline src/models/Bookmark.model.ts:26]: ref: "User"

File: src/repositories/bookmarks.repository.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import { BookmarkModel } from "../models/Bookmark.model";
+
+export const listBookmarks = async () => {
+  const result = await BookmarkModel.find();
+  return result;
+};
+
+export const addOne = async (tweetId: string, ownerId: string) => {
+  const result = new BookmarkModel({
+    tweetId: new ObjectId(tweetId),
```
Comment (alexandrtovmach) [inline src/repositories/bookmarks.repository.ts]:
```suggestion
    tweetId,
```

File: src/repositories/bookmarks.repository.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    ownerId: new ObjectId(ownerId),
```
Comment (alexandrtovmach) [inline src/repositories/bookmarks.repository.ts]:
```suggestion
   ownerId,
```

File: src/repositories/bookmarks.repository.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  });
+  return result.save();
+};
+
+export const deleteOne = async (id: string) => {
+  const result = await BookmarkModel.deleteOne({ _id: new ObjectId(id) });
```
Comment (alexandrtovmach) [inline src/repositories/bookmarks.repository.ts]: use deleteById

**Review by alexandrtovmach — CHANGES_REQUESTED** [review]

**Review by alexandrtovmach — COMMENTED** [review]
we already talked, ISOString instead of Date

File: src/models/Bookmark.model.ts [hunk 13]
```diff
@@ -0,0 +1,35 @@
+import { ObjectId } from "mongodb";
+import mongoose, { model, Schema, Document } from "mongoose";
+
+export interface Bookmark extends Document {
+  tweetId: ObjectId;
+  ownerId: ObjectId;
+  createdAt: string;
+  updatedAt: string | null;
```
Comment (alexandrtovmach) [inline src/models/Bookmark.model.ts]:
```suggestion
  updatedAt?: string;
```

File: src/models/Bookmark.model.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    ref: "User",
+  },
+
+  createdAt: {
+    type: String,
+    required: true,
+  },
+
+  updatedAt: {
+    type: String || null,
```
Comment (alexandrtovmach) [inline src/models/Bookmark.model.ts]:
```suggestion
    type: String,
```

File: src/repositories/bookmarks.repository.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+export const listBookmarks = async () => {
+  const result = await BookmarkModel.find();
+  return result;
+};
+
+export const addOne = async (tweetId: string, ownerId: string) => {
+  const result = new BookmarkModel({
+    tweetId: tweetId,
+    ownerId: ownerId,
+    createdAt: new Date().toISOString(),
```
Comment (alexandrtovmach) [inline src/repositories/bookmarks.repository.ts:11]:
```suggestion
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
```

**Review by alexandrtovmach — CHANGES_REQUESTED** [review]

File: src/repositories/bookmarks.repository.ts [hunk 16]
```diff
@@ -0,0 +1,23 @@
+import { BookmarkModel } from "../models/Bookmark.model";
+
+export const listBookmarks = async () => {
+  const result = await BookmarkModel.find();
+  return result;
+};
```
Comment (VitaliiDenysiuk) [inline src/repositories/bookmarks.repository.ts]:
```suggestion
export const listBookmarks =  () => {
  return BookmarkModel.find();
};
```

File: src/repositories/bookmarks.repository.ts [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+
+export const addOne = async (tweetId: string, ownerId: string) => {
```
Comment (VitaliiDenysiuk) [inline src/repositories/bookmarks.repository.ts]:
```suggestion
export const addOne =  (tweetId: string, ownerId: string) => {
```
await absent - async is not needed

File: src/repositories/bookmarks.repository.ts [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    createdAt: new Date().toISOString(),
+    updatedAt: new Date().toISOString(),
+  });
+  return result.save();
+};
+
+export const deleteById = async (id: string) => {
+  const result = await BookmarkModel.deleteOne({ _id: id });
+  return result;
+};
```
Comment (VitaliiDenysiuk) [inline src/repositories/bookmarks.repository.ts]:
```suggestion
export const deleteById = (id: string) => {
  return BookmarkModel.deleteOne({ _id: id });
};
```

**Review by VitaliiDenysiuk — CHANGES_REQUESTED** [review]
fix some remarks

**Review by alexandrtovmach — APPROVED** [review]

**Review by VitaliiDenysiuk — APPROVED** [review]

**Review by VitaliiDenysiuk — APPROVED** [review]

**Pull request merged**